rpc_util.go

```diff
@@ -307,40 +309,45 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
 			outPayload.Data = b
 			outPayload.Length = len(b)
 		}
+
 		if cp != nil {
+			// pre-alloc place
+			cbuf.Write(make([]byte, 5))
 			if err := cp.Do(cbuf, b); err != nil {
 				return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
 			}
 			b = cbuf.Bytes()
+			length = uint(len(b) - 5)
+		} else {
+			length = uint(len(b))
 		}
-		length = uint(len(b))
 	}
+
 	if length > math.MaxUint32 {
```

Comment:
That's not an option -- we must follow the protocol, which is 4 bytes to encode the length.

My point is about the type of the variables.  Checking whether a uint is > MaxUint32 will only work on platforms where a uint is 64 bits.

tl;dr: please make length a uint64.